Std/Data/BitVec/Basic.lean

```diff
@@ -0,0 +1,185 @@
+/-
+Copyright (c) 2022 by the authors listed in the file AUTHORS and their
+institutional affiliations. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joe Hendrix, Wojciech Nawrocki, Leonardo de Moura
+-/
+namespace Std
+/-!
+We define bitvectors. We choose the `Fin` representation over others for its relative efficiency
+(Lean has special support for `Nat`), alignment with `UIntXY` types which are also represented
+with `Fin`, and the fact that bitwise operations on `Fin` are already defined. Some other possible
+representations are `List Bool`, `{ l : List Bool // l.length = w}`, `Fin w → Bool`.
+
+We also define many bitvector operations from the
+[`QF_BV` logic](https://smtlib.cs.uiowa.edu/logics-all.shtml#QF_BV).
+of SMT-LIBv2.
+
+TODO(Leo): match the interface to what SMT-LIB provides
+-/
+
+/--
+A bitvector of the specified width. This is represented as the underlying `Nat` number
+in both the runtime and the kernel, inheriting all the special support for `Nat`.
+-/
+structure BitVec (w : Nat) where
+  /-- Bitvector is represented as a `Fin`. -/
+  val : Fin (2^w)
+  deriving DecidableEq
+
+namespace BitVec
+
+theorem pow_two_gt_zero (w : Nat) : 2^w > 0 := by
+  apply Nat.pos_pow_of_pos; decide
+
+/-- The BitVector `i mod 2^n`. -/
+protected def ofNat (n : Nat) (i : Nat) : BitVec n :=
+  { val := Fin.ofNat' i (pow_two_gt_zero _) }
+
+/-- Given a bitvector `a`, return the underlying `Nat`. -/
+protected def toNat (a : BitVec n) : Nat :=
+  a.val.val
+
+/-- Return a bitvector `0` of size `n`.  -/
+protected def zero (n : Nat) : BitVec n :=
+  BitVec.ofNat n 0
+
+instance : Inhabited (BitVec n) where
+  default := BitVec.zero n
+
+instance : OfNat (BitVec n) i where
+  ofNat := BitVec.ofNat n i
+
+/--
+Notation for bit vector literals. `i#n` is a shorthand for `BitVec.ofNat n i`.
+TODO: is this a good notation? Do we need it? We can also write `(i : BitVec n)`
+-/
+syntax:max num noWs "#" noWs num : term
```

Comment:
```suggestion
scoped syntax:max num noWs "#" noWs num : term
```

Reply:
If we keep the notation, probably best to use it below, to make it more familiar to API users.

(Thumbs up and I'll make this change.)

Reply:
It seems narrowly useful if it only works for literals. I'd say we don't want it.

Reply:
It makes proof goals way more readable. Specially the ones coming from software/hardware verification.

Reply:
I agree with Scott, the only thing this does is add a type ascription so it seems more suited as either `#n` as a notation for `Bitvec n` (so `(5 : #3)` works), or letting downstream projects define the notation themselves as a local notation (it's a one-liner). I don't find the particular syntax very descriptive and would prefer the definition be moved closer to the use for readability.

Reply:
Happy to change the notation, but it would be great to have something very compact and working "out of the box" for people working on software/hardware verification.

Reply:
My gut feeling is that writing the arity all the time will get old quick, especially in cases where the arity is mostly fixed (e.g. everything is on 32 bit words). For sure frequently varied arity definitely comes up too, especially in hardware verification, but I don't think it is the majority case, and inventing a new domain specific notation for type ascription seems like it would lead to more confusion and not less, unless this is a very standard notation in the trade (and my impression is that it is not). It might be good to do a survey of what VHDLs and such do for this situation.

Reply:
I think I am slightly in favor of the notation.  This is probably orthogonal, but it'd be neat if there was a way to have access to the number of digits in a hex or binary literal (e.g., if we knew `0b01010` was a BitVec, then it would default to having 5 bits).  Similar with hex digits having a multiple of the input `0x0010` is a `BitVec 16`.  Cryptol does this, but also is predominantly about bitvector programming.